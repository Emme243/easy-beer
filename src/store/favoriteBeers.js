export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    addBeerToFavorites({ commit }, beer) {
      commit('addBeerToFavorites', beer);
    },
    removeBeerFromFavorites({ commit }, beerId) {
      commit('removeBeerFromFavorites', beerId);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
};
