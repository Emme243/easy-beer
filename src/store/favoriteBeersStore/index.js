export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    handleBeerInFavorites({ commit, getters }, beer) {
      const isBeerInFavorites = getters.isBeerInFavorites(beer.id);
      if (isBeerInFavorites) commit('removeBeerFromFavorites', beer.id);
      else commit('addBeerToFavorites', beer);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
};
