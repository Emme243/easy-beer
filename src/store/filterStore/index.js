import { abvRangeStore } from '@/store/filterStore/abvRangeStore';
import { brewedDateStore } from '@/store/filterStore/brewedDateStore';
import { paginationStore } from '@/store/filterStore/paginationStore';
import { searchQueryStore } from '@/store/filterStore/searchQueryStore';

export const filterStore = {
  namespaced: true,
  modules: {
    abvRange: abvRangeStore,
    brewedDate: brewedDateStore,
    page: paginationStore,
    searchQuery: searchQueryStore,
  },
  state: () => ({}),
  getters: {
    filterStore(_, getters) {
      return {
        minAbvValue: getters.minAbvValue,
        maxAbvValue: getters.maxAbvValue,
        initialBrewedMonth: getters.initialBrewedMonth,
        finalBrewedMonth: getters.finalBrewedMonth,
        page: getters.page,
        searchQuery: getters.searchQuery,
      };
    },
    areFiltersWithDefaultValue(_, getters) {
      return (
        getters.isAbvRangeWithDefaultValue &&
        !getters.initialBrewedMonth &&
        !getters.finalBrewedMonth &&
        !getters.searchQuery
      );
    },
    numberOfFiltersApplied(_, getters) {
      let numberOfFiltersApplied = 0;
      if (!getters.isAbvRangeWithDefaultValue) numberOfFiltersApplied++;
      if (getters.initialBrewedMonth || getters.finalBrewedMonth) numberOfFiltersApplied++;
      return numberOfFiltersApplied;
    },
  },
  mutations: {},
  actions: {
    setAllFilters({ dispatch }, filterQuery) {
      dispatch('setPage', filterQuery.page);
      dispatch('setMinAbvValue', filterQuery.minAbvValue);
      dispatch('setMaxAbvValue', filterQuery.maxAbvValue);
      dispatch('setInitialBrewedMonth', filterQuery.initialBrewedMonth);
      dispatch('setFinalBrewedMonth', filterQuery.finalBrewedMonth);
      dispatch('setSearchQuery', filterQuery.searchQuery);
    },
    resetAllFilters({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('setSearchQuery', '');
      dispatch('resetPage');
    },
    resetFiltersInFilterModal({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('resetPage');
    },
  },
};
