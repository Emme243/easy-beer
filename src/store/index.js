import { favoriteBeersStore } from '@/store/favoriteBeersStore';
import { filterStore } from '@/store/filterStore';

import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    filter: filterStore,
    favoriteBeers: favoriteBeersStore,
  },
});
