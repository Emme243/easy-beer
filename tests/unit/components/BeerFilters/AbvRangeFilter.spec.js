import AbvRangeFilter from '@/components/BeerFilters/AbvRangeFilter';
import store from '@/store';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuetify from 'vuetify';

describe('Pruebas en el componente AbvRangeFilter.vue', function () {
  let vuetify;
  const localVue = createLocalVue();

  const mountFunction = options => {
    return shallowMount(AbvRangeFilter, {
      localVue,
      vuetify,
      store,
      ...options,
    });
  };

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  const resetButtonSelector = '#reset-abv-range-btn';
  const abvRangeSlideSelector = '[data-test-id="abv-range-slider"]';
  const abvInfoMessageSelector = '[data-test-id="abv-info-message"]';

  const abvInfoMessage = (minAbvValue, maxAbvValue) =>
    `Se mostrarán cervezas con % de alcohol entre el ${minAbvValue}% y el ${maxAbvValue}%`;

  it('Debe hacer match con el snapshot', function () {
    const wrapper = mountFunction();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('El botón de reseteo debe llamar al dispatch resetAbvRange cuando se hace click', async function () {
    store.dispatch = jest.fn();
    const wrapper = mountFunction();
    const resetButton = wrapper.find(resetButtonSelector);
    await resetButton.trigger('click');
    expect(store.dispatch).toHaveBeenCalledWith('filter/resetAbvRange');
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });

  it('Debe llamar al dispatch resetPage, setMinAbvValue y setMaxAbvValue cuando se hace un cambio en los rangos de abv', async function () {
    const newMinAbvValue = 5;
    const newMaxAbvValue = 10;
    store.dispatch = jest.fn();
    const wrapper = mountFunction();
    const abvRangeSlide = wrapper.find(abvRangeSlideSelector);
    expect(abvRangeSlide.exists()).toBe(true);
    abvRangeSlide.vm.$emit('change', [newMinAbvValue, newMaxAbvValue]);
    await wrapper.vm.$nextTick();
    expect(store.dispatch).toHaveBeenCalledWith('filter/resetPage');
    expect(store.dispatch).toHaveBeenCalledWith('filter/setMinAbvValue', newMinAbvValue);
    expect(store.dispatch).toHaveBeenCalledWith('filter/setMaxAbvValue', newMaxAbvValue);
    expect(store.dispatch).toHaveBeenCalledTimes(3);
  });

  it('El abv info message debe mostrar el mensaje correcto con los valores de abv predeterminados (2 y 60)', function () {
    const wrapper = mountFunction();
    const abvInfoMessageElement = wrapper.find(abvInfoMessageSelector);
    const { minAbvValue, maxAbvValue } = wrapper.vm;
    expect(abvInfoMessageElement.text()).toBe(abvInfoMessage(minAbvValue, maxAbvValue));
  });

  it('El abv info message debe mostrar el mensaje correcto con los valores de abv modificados', function () {
    const minAbvValue = 5;
    const maxAbvValue = 10;
    const wrapper = mountFunction({
      computed: {
        minAbvValue() {
          return minAbvValue;
        },
        maxAbvValue() {
          return maxAbvValue;
        },
      },
    });
    const abvInfoMessageElement = wrapper.find(abvInfoMessageSelector);
    expect(abvInfoMessageElement.text()).toBe(abvInfoMessage(minAbvValue, maxAbvValue));
  });
});
