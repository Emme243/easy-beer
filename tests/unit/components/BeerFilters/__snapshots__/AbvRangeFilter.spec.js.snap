// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente AbvRangeFilter.vue Debe hacer match con el snapshot 1`] = `
<div>
  <div class="d-flex align-center">
    <h3 class="text-h6 mr-4">Volumen de alcohol</h3> <button id="reset-abv-range-btn" style="display: none;">
      <v-icon-stub small="">fa fa-arrow-rotate-left</v-icon-stub>
    </button>
  </div> <span>Cantidad de alcohol en porcentaje (%) que una cerveza tiene.</span>
  <v-range-slider-stub errorcount="1" errormessages="" messages="" rules="" successmessages="" value="2,60" backgroundcolor="" dense="true" hidedetails="true" persistenthint="true" loaderheight="2" max="60" min="2" step="1" thumblabel="true" thumbsize="32" ticklabels="" ticksize="2" data-test-id="abv-range-slider"></v-range-slider-stub> <span data-test-id="abv-info-message" class="font-weight-bold">
    Se mostrar√°n cervezas con % de alcohol entre el 2% y el 60%
  </span>
</div>
`;
